library/src/main/kotlin/com/livefront/gsonkotlinadapter/KoltinReflectiveTypeAdapterFactory.kt

```diff
@@ -107,6 +122,10 @@ class KotlinReflectiveTypeAdapterFactory private constructor() : TypeAdapterFact
             }
             reader.endObject()
 
+            // Add all default values if the JSON did not include it
+            constructorParameterDefaultsMap.map { (parameter, data) ->
+                constructorParams.putIfAbsent(parameter, data)
+            }
             return primaryConstructor.callBy(constructorParams)
```

Comment:
I think making use of `Map.plus` here would be better than using `putIfAbsent`, since `putIfAbsent` interacts really strangely with `null` values since its a `Java` method:

```kotlin
    @Test
    fun david() {
        val alpha = mutableMapOf<String, Any?>()
        alpha["a"] = null
        alpha.putIfAbsent("a", "test")
        assertEquals(mapOf("a" to "test"), alpha)
    }
```

In this case, it's impossible that a `null` value would be replaced by a non-`null` value incorrectly, but it could be a really weird source of bugs in the future.


```suggestion
            return primaryConstructor.callBy(constructorParameterDefaultsMap + constructorParams)
```

Reply:
Related to the above, the following works, and doesn't throw an exception with "declares multiple JSON fields named":

```kotlin
    data class David(
        val string: String?
    )

    @Test
    fun david() {
        val expectedData = David("string")
        val actualData = gson.fromJson<David>(
            """
                {
                    "string": null,
                    "string": null,
                    "string": null,
                    "string": null,
                    "string": "string"
                }
            """.trimIndent()
        )
        assertEquals(expectedData, actualData)
    }
```

Reply:
Hmm, okay using `Map.plus` does change one case: Should the following test pass? (It fails with `gson` instead of `gsonWithDefaultPrimitiveValuesEnabled`)

```kotlin
    data class David(
        val integer: Int
    )

    @Test
    fun david() {
        val expectedData = David(0)
        val actualData = gsonWithDefaultPrimitiveValuesEnabled.fromJson<David>(
            """
                {
                    "integer": null
                }
            """.trimIndent()
        )
        assertEquals(expectedData, actualData)
    }
```